addons/mail/static/src/rtc/voice_recorder.js

```diff
@@ -0,0 +1,136 @@
+        if (!this.microphone) {
+            try {
+                this.microphone = await browser.navigator.mediaDevices.getUserMedia({
+                    audio: this.userSettings.audioConstraints,
+                });
+            } catch {
+                this.notification.add(
+                    sprintf(_t('"%(hostname)s" needs to access your microphone'), {
+                        hostname: window.location.host,
+                    }),
+                    { type: "warning" }
+                );
+                this.state.isActionPending = false;
+                return;
+            }
+        }
+        this.state.elapsed = "00 : 00";
+        this.state.recording = true;
+        let totalSeconds = 0;
+        this.state.intervalId = setInterval(() => {
+            ++totalSeconds;
+            const second = totalSeconds % 60;
+            const minute = Math.floor(totalSeconds / 60) % 60;
+            this.state.elapsed =
+                (minute < 10 ? "0" + minute : minute) +
+                " : " +
+                (second < 10 ? "0" + second : second);
+            if (totalSeconds == 60) {
+                this.notification.add(
+                    _t("The voice message duration should be up to one minute."),
+                    { type: "warning" }
+                );
+                this.stopRecording();
+            }
+        }, 1000);
+        const mimeType = this.isSafari ? "audio/mp4" : "audio/webm";
+        this.recorder = new MediaRecorder(this.microphone, { mimeType });
+        this.startTimeStamp = Date.now();
+        this.recorder.ondataavailable = (e) => {
+            if (this.unmount) {
+                return;
+            }
+            const blob = new Blob([e.data], {
+                type: this.recorder.mimeType || mimeType,
+            });
+            this.duration = e.timecode - this.startTimeStamp;
+            const file = new File([blob], this.filename(), {
+                type: "voice/message",
+            });
+            this.props.attachmentUploader.uploadFile(file);
+            this.recorder = null;
+            this.props.composer.hasVoice = true;
+        };
+        this.recorder.start(24 * 60 * 60 * 1000);
+        this.state.isActionPending = false;
+    }
+
+    stopRecording() {
+        if (this.recorder?.state === "recording") {
+            this.recorder.stop();
+            clearInterval(this.state.intervalId);
+            this.state.intervalId = false;
+        }
+        this.microphone?.getTracks().forEach((track) => track.stop());
+        this.microphone = null;
+        this.state.recording = false;
+        if (!this.unmount) {
+            this.props.onchangeRecording();
+        }
+    }
+
+    onClick() {
+        if (this.state.recording) {
+            this.stopRecording();
+        } else {
+            this.startRecording();
+        }
+    }
+
+    filename() {
+        const extension = this.isSafari ? "audio/mp4" : "audio/webm";
```

Comment:
I don't think that it's a good extension but the fact that you use a custom mimetype will also prevent the files from working with some audio readers external to odoo.

Is it intentional that the compatibility with other readers is broken? 
I don't know much about the flow of mail messages but I thought that there could be a case where the attachments are also visible from mails that are sent from odoo, in which case it is unlikely that these files will be readable.